eth_block.go

```diff
@@ -0,0 +1,391 @@
+	}
+
+	// Process the found eth-tx objects
+	ethTxNodes, ethTxTrieNodes := processTransactions(decodedBlock.Transactions(),
+		decodedBlock.Header().TxHash[:])
+
+	return ethBlock, ethTxNodes, ethTxTrieNodes, nil
+}
+
+// FromBlockJSON takes the output of an ethereum client JSON API
+// (i.e. parity or geth) and returns a set of IPLD nodes.
+func FromBlockJSON(r io.Reader) (*EthBlock, []*EthTx, []*EthTxTrie, error) {
+	var obj objJSONBlock
+	dec := json.NewDecoder(r)
+	err := dec.Decode(&obj)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	headerRawData := getRLP(obj.Result.Header)
+	ethBlock := &EthBlock{
+		Header:  &obj.Result.Header,
+		cid:     rawdataToCid(MEthBlock, headerRawData),
+		rawdata: headerRawData,
+	}
+
+	// Process the found eth-tx objects
+	ethTxNodes, ethTxTrieNodes := processTransactions(obj.Result.Transactions,
+		obj.Result.Header.TxHash[:])
+
+	return ethBlock, ethTxNodes, ethTxTrieNodes, nil
+}
+
+// processTransactions will take the found transactions in a parsed block body
+// to return IPLD node slices for eth-tx and eth-tx-trie
+func processTransactions(txs []*types.Transaction, expectedTxRoot []byte) ([]*EthTx, []*EthTxTrie) {
+	var ethTxNodes []*EthTx
+	transactionTrie := newTxTrie()
+
+	for idx, tx := range txs {
+		ethTx := NewTx(tx) // Will panic if it finds an error while parsing a tx
+		ethTxNodes = append(ethTxNodes, ethTx)
+		transactionTrie.addTx(idx, ethTx.RawData())
+	}
+
+	ethTxTrieNodes := transactionTrie.getNodes()
+	if !bytes.Equal(transactionTrie.rootHash(), expectedTxRoot) {
+		panic("Wrong transaction hash computed!")
+	}
+
+	return ethTxNodes, ethTxTrieNodes
+}
+
+/*
+  OUTPUT
+*/
+
+// DecodeEthBlock takes a cid and its raw binary data
+// from IPFS and returns an EthBlock object for further processing.
+func DecodeEthBlock(c *cid.Cid, b []byte) (*EthBlock, error) {
+	var h types.Header
+	err := rlp.Decode(bytes.NewReader(b), &h)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EthBlock{
+		Header:  &h,
+		cid:     c,
+		rawdata: b,
+	}, nil
+}
+
+/*
+  Block INTERFACE
+*/
+
+// RawData returns the binary of the RLP encode of the block header.
+func (b *EthBlock) RawData() []byte {
+	return b.rawdata
+}
+
+// Cid returns the cid of the block header.
+func (b *EthBlock) Cid() *cid.Cid {
+	return b.cid
+}
+
+// String is a helper for output
+func (b *EthBlock) String() string {
+	return fmt.Sprintf("<EthBlock %s>", b.cid)
+}
+
+// Loggable returns a map the type of IPLD Link.
+func (b *EthBlock) Loggable() map[string]interface{} {
+	return map[string]interface{}{
+		"type": "eth-block",
+	}
+}
+
+/*
+  Node INTERFACE
+*/
+
+// Resolve resolves a path through this node, stopping at any link boundary
+// and returning the object found as well as the remaining path to traverse
+func (b *EthBlock) Resolve(p []string) (interface{}, []string, error) {
+	if len(p) == 0 {
+		return b, nil, nil
+	}
+
+	// Sanity check for terminal paths
+	nonLinkPaths := map[string]interface{}{
```

Comment:
Maybe better (faster) to just use two switches. That is, a switch for links (where `len(p)` can be `> 1`), then a check for `len(p) == 1`, then a switch for non-links.